### PR TITLE
[muP] Rework

### PR DESCRIPTION
**Pull request by lintangsutawika**

### CODE REVIEW

File: megatron/model/init_functions.py [hunk 1]
```diff
@@ -16,41 +16,22 @@
...
 
-try:
-    import mup
-except ImportError:
-    pass
 
-
-def init_method_normal(sigma, use_mup_outer=False, mup_init_scale=1.0):
+def init_method_normal(sigma):
     """Init method based on N(0, sigma)."""
```
Comment (ndey96) [inline megatron/model/init_functions.py:21]: shouldnt it be N(0, sigma^2)?

File: megatron/model/init_functions.py [hunk 2]
```diff
@@ -103,105 +84,91 @@ def init_(tensor):
...
-            mup.init.xavier_uniform_(tensor)
+    def init_(tensor, mup_m_width=mup_m_width):
+        init_weight = torch.nn.init.xavier_uniform_(tensor)
+        if mup_m_width != 1:
             with torch.no_grad():
-                tensor.mul_(mup_init_scale)
-            return tensor
-        else:
-            return torch.nn.init.xavier_uniform_(tensor)
+                init_weight.div_(mup_m_width)
```
Comment (ndey96) [inline megatron/model/init_functions.py:95]: Bug: Dividing a random variable by `mup_m_width` will scale its variance by `1/(mup_m_width**2)`. Instead you want to divide by `sqrt(mup_m_width)`

File: megatron/model/init_functions.py [hunk 3]
```diff
@@ -103,105 +84,91 @@ def init_(tensor):
...
-            mup.init.xavier_normal_(tensor)
+    def init_(tensor, mup_m_width=mup_m_width):
+        init_weight = torch.nn.init.xavier_normal_(tensor)
+        if mup_m_width != 1:
             with torch.no_grad():
-                tensor.mul_(mup_init_scale)
-            return tensor
-        else:
-            return torch.nn.init.xavier_normal_(tensor)
+                init_weight.div_(mup_m_width)
```
Comment (ndey96) [inline megatron/model/init_functions.py:109]: Potentially same bug as above

File: megatron/model/init_functions.py [hunk 4]
```diff
@@ -103,105 +84,91 @@ def init_(tensor):
...
-            mup.init.normal_(tensor, mean=0.0, std=std)
+    def init_(tensor, mup_m_width=mup_m_width):
+        init_weight = torch.nn.init.normal_(tensor, mean=0.0, std=std)
+        if mup_m_width != 1:
             with torch.no_grad():
-                tensor.mul_(mup_init_scale)
-            return tensor
-        else:
-            return torch.nn.init.normal_(tensor, mean=0.0, std=std)
+                init_weight.div_(mup_m_width)
```
Comment (ndey96) [inline megatron/model/init_functions.py:124]: Potentially same bug as above

File: megatron/model/init_functions.py [hunk 5]
```diff
@@ -103,105 +84,91 @@ def init_(tensor):
...
+    def init_(tensor, mup_m_width=mup_m_width):
+        init_weight = torch.nn.init.normal_(tensor, mean=0.0, std=std)
+        if mup_m_width != 1:
             with torch.no_grad():
-                tensor.mul_(mup_init_scale)
-            return tensor
-        else:
-            return torch.nn.init.normal_(tensor, mean=0.0, std=std)
-
+                init_weight.div_(mup_m_width)
```
Comment (ndey96) [inline megatron/model/init_functions.py:137]: Potentially same bug as above

File: megatron/model/init_functions.py [hunk 6]
```diff
@@ -103,105 +84,91 @@ def init_(tensor):
...
-            import mup
-        except ModuleNotFoundError:
-            print("Please install mup https://github.com/microsoft/mup")
-            raise Exception
-
     def _get(name):
         if name == "normal":
             return init_method_normal(
-                args.init_method_std, args.use_mup, args.mup_init_scale
+                sigma=args.init_method_std/math.sqrt(args.mup_m_width)
```
Comment (ndey96) [inline megatron/model/init_functions.py:148]: if you are going to define the muP init_std adjustment here then you need to remove all the `init_weight.div_` calls so you dont double adjust

File: megatron/model/transformer.py [hunk 7]
```diff
@@ -971,6 +971,9 @@ def parallel_lm_logits(input_, word_embeddings_weight, parallel_output, bias=Non
     else:
         logits_parallel = F.linear(input_parallel, word_embeddings_weight, bias)
 
+    if args is not None and args.use_mup:
+        logits_parallel /= args.mup_m_width
```
Comment (ndey96) [inline megatron/model/transformer.py:975]: you may also want to multiply by some tunable scalar value here (like we did in BLTM)

File: megatron/model/word_embeddings.py [hunk 8]
```diff
@@ -155,9 +154,10 @@ def forward(self, input_ids, position_ids, tokentype_ids=None):
         # Dropout.
         embeddings = self.embedding_dropout(embeddings)
 
+        # Y_emb = m_emb * embed(X)
         if self.use_mup:
             with torch.no_grad():
-                embeddings.mul_(self.mup_embedding_mult)
+                embeddings = torch.mul(embeddings, self.mup_m_emb)
```
Comment (ndey96) [inline megatron/model/word_embeddings.py:160]: seems correct, where mup_m_emb is a tunable scalar

File: megatron/mpu/layers.py [hunk 9]
```diff
@@ -547,8 +548,10 @@ def set_parallel_output(self, value: bool):
         )  # if gather_output is True, parallel output is False, so we set the opposite
 
     def forward(self, input_):
+
+        # Y_logits = W_unembed * X / m_width
         if self.use_mup and self.mup_rescale_parameters:
-            input_ /= self.width_mult()
+            input_ /= self.m_width
```
Comment (ndey96) [inline megatron/mpu/layers.py:554]: This is unnecessary and should be removed. The only adjustments that need to be made are to the LR and init_std. This will apply an adjustment twice. I think its safe to remove all muP stuff from this file

File: megatron/mup_substitute.py [hunk 10]
```diff
@@ -39,31 +41,51 @@ def _get_coord_data(
 ):
     df = []
 
+    def word_embedding_coord_check_hook(module, input, output):
+        with torch.no_grad():
+            word_embedding_act_abs_mean_list.append(output.abs().mean().item())
+
+    word_embedding_act_abs_mean_list = []
```
Comment (ndey96) [inline megatron/mup_substitute.py:48]: i think this needs to be defined above the hook?

File: megatron/mup_substitute.py [hunk 11]
```diff
@@ -39,31 +41,51 @@ def _get_coord_data(
 ):
```
Comment (ndey96) [inline megatron/mup_substitute.py:41]: To implement the coord check heres what I did:

`megatron/training.py::train()`
image

`megatron/training.py::training_log()`
image

Then to run a coord check I just configure a 10 step training run where these scalars are being logged. To make the coord check graphs you can just parse your tensorboard logs 
cc @lintangsutawika @Quentin-Anthony

Comment (ndey96) [reply megatron/mup_substitute.py:41]: Then in `arguments.py` I add this to enable coord check:
image

File: megatron/learning_rates.py [hunk 12]
```diff
@@ -95,8 +97,8 @@ def step(self, step_num=None):
         self.num_iters = step_num
         new_lr = self.get_lr()
         for group in self.optimizer.param_groups:
-            if self.use_mup and "width_mult" in group:
-                group["lr"] = new_lr / group["width_mult"]
+            if self.use_mup and ("lr_adjust" in group) and group["lr_adjust"] is True:
+                group["lr"] = new_lr / self.mup_m_width
```
Comment (ndey96) [inline megatron/learning_rates.py:101]: 👍🏾

File: megatron/model/init_functions.py (hunk 6 above)
Comment (lintangsutawika) [reply megatron/model/init_functions.py]: I'm split on what the best interface would be. For `init_method_normal` and `scaled_init_method_normal` I opted to not scale in the function but for other init functions I did.

File: megatron/model/transformer.py (hunk 7 above)
Comment (lintangsutawika) [reply megatron/model/transformer.py]: I thought that `mup_m_width` was the tunable scalar for the logits here?

File: megatron/mpu/layers.py (hunk 9 above)
Comment (lintangsutawika) [reply megatron/mpu/layers.py:554]: I intended to implement the formula `Y_logits = W_unembed * X / m_width` from this table, specifically `output_logits_multiplier`
image

File: megatron/model/transformer.py (hunk 7 above)
Comment (lintangsutawika) [reply megatron/model/transformer.py]: Is this the part where Y_logits = W_unemb X/m_width?

File: megatron/mpu/layers.py (hunk 9 above)
Comment (lintangsutawika) [reply megatron/mpu/layers.py:554]: Nvm, I realized where I was wrong. `Output logits multiplier` refers to the output embedding layer and not the output of the intermediate layers. I've removed the scaling in `layers.py`

File: megatron/mpu/layers.py [hunk 13]
```diff
@@ -735,7 +739,7 @@ def set_parallel_output(self, parallel_output: bool):
 
     def forward(self, input_):
         if self.use_mup and self.mup_rescale_parameters:
-            input_ /= self.width_mult()
+            input_ /= self.m_width
```
Comment (ndey96) [inline megatron/mpu/layers.py]: also redundant

Comment (lintangsutawika) [inline megatron/mpu/layers.py]: this should also be removed.

File: megatron/model/transformer.py [hunk 14]
```diff
@@ -971,6 +969,9 @@ def parallel_lm_logits(input_, word_embeddings_weight, parallel_output, bias=Non
     else:
         logits_parallel = F.linear(input_parallel, word_embeddings_weight, bias)
 
+    if args is not None and args.use_mup:
+        logits_parallel /= args.mup_width_multiplier
+
```
Comment (lintangsutawika) [inline megatron/model/transformer.py:974]:
```
logits_parallel /= args.mup_width_multiplier
logits_parallel *= args.mup_output_multiplier 
```

File: megatron/model/init_functions.py [hunk 15]
```diff
@@ -103,105 +84,91 @@ def init_(tensor):
...
-            import mup
-        except ModuleNotFoundError:
-            print("Please install mup https://github.com/microsoft/mup")
-            raise Exception
-
     def _get(name):
         if name == "normal":
             return init_method_normal(
-                args.init_method_std, args.use_mup, args.mup_init_scale
+                sigma=args.init_method_std/math.sqrt(args.mup_width_multiplier)
```
Comment (jthestness) [inline megatron/model/init_functions.py:148]: During our call, we noted that this leads to a bug: Since the width multiplier is applied to all layers, this doesn't allow the embedding layer to be initialized differently from the transformer backbone layers (precisely: muP prescribes that layers who's input and output dimensions both scale with width need to have a sqrt(width) multiplying factor).

Comment (jthestness) [reply megatron/model/init_functions.py:148]: Would recommend refactoring this code: Remove the muP width multiplier completely from the initialization methods code, and only take the initializer parameters in here (e.g., standard deviation). Then, when the initializers are used from various layers, adjust the initializer based on that particular layer's muP width adjustment requirements.

**Review by rezaarefi — APPROVED** [review]
